src/Microsoft.Tye.Hosting/Infrastructure/IngressHostMatcherPolicy.cs

```diff
@@ -0,0 +1,209 @@
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            });
+        }
+
+        public Task ApplyAsync(HttpContext httpContext, CandidateSet candidates)
+        {
+            if (httpContext == null)
+            {
+                throw new ArgumentNullException(nameof(httpContext));
+            }
+
+            if (candidates == null)
+            {
+                throw new ArgumentNullException(nameof(candidates));
+            }
+
+            for (var i = 0; i < candidates.Count; i++)
+            {
+                if (!candidates.IsValidCandidate(i))
+                {
+                    continue;
+                }
+
+                var hosts = candidates[i].Endpoint.Metadata.GetMetadata<IngressHostMetadata>()?.Hosts;
+                if (hosts == null || hosts.Count == 0)
+                {
+                    // Can match any host.
+                    continue;
+                }
+
+                var matched = false;
+                var (requestHost, requestPort) = GetHostAndPort(httpContext);
+                for (var j = 0; j < hosts.Count; j++)
+                {
+                    var host = hosts[j].AsSpan();
+                    var port = ReadOnlySpan<char>.Empty;
+
+                    // Split into host and port
+                    var pivot = host.IndexOf(':');
+                    if (pivot >= 0)
+                    {
+                        port = host.Slice(pivot + 1);
+                        host = host.Slice(0, pivot);
+                    }
+
+                    if (host == null || MemoryExtensions.Equals(host, WildcardHost, StringComparison.OrdinalIgnoreCase))
+                    {
+                        // Can match any host
+                    }
+                    else if (
+                        host.StartsWith(WildcardPrefix) &&
+
+                        // Note that we only slice of the `*`. We want to match the leading `.` also.
+                        MemoryExtensions.EndsWith(requestHost, host.Slice(WildcardHost.Length), StringComparison.OrdinalIgnoreCase))
+                    {
+                        // Matches a suffix wildcard.
+                    }
+                    else if (MemoryExtensions.Equals(requestHost, host, StringComparison.OrdinalIgnoreCase))
+                    {
+                        // Matches exactly
+                    }
+                    else
+                    {
+                        // If we get here then the host doesn't match.
+                        continue;
+                    }
+
+                    if (MemoryExtensions.Equals(port, WildcardHost, StringComparison.OrdinalIgnoreCase))
+                    {
+                        // Port is a wildcard, we allow any port.
+                    }
+                    else if (port.Length > 0 && (!int.TryParse(port, out var parsed) || parsed != requestPort))
```

Comment:
Just some thoughts in my head: this all works when you have multiple replicas per app. The port selection is based on the exposed port, not the replica ports, right?

Reply:
Everything here is based on replica port.